From bda664ba3405100f7ccba8b3958c14a893830ab7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lo=C3=AFc=20Molinari?= <loic.molinari@collabora.com>
Date: Thu, 3 Sep 2026 16:56:04 +0200
Subject: [PATCH 1/3] panfrost: Log AFBC-P messages at debug level (not warn)
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Use mesa_logd() directly instead of trying to copy perf_debug() which
is meant to support GL debug output. The warnings were also cluttering
CI output when AFBC-P is enabled.

Signed-off-by: Loïc Molinari <loic.molinari@collabora.com>
---
 src/gallium/drivers/panfrost/pan_resource.c | 31 ++++++++++-----------
 src/gallium/drivers/panfrost/pan_screen.h   |  7 -----
 2 files changed, 14 insertions(+), 24 deletions(-)

diff --git a/src/gallium/drivers/panfrost/pan_resource.c b/src/gallium/drivers/panfrost/pan_resource.c
index 178b7b8a5624..dab9f3681f62 100644
--- a/src/gallium/drivers/panfrost/pan_resource.c
+++ b/src/gallium/drivers/panfrost/pan_resource.c
@@ -1961,12 +1961,11 @@ pan_resource_afbcp_get_payload_sizes(struct panfrost_context *ctx,
 {
    PAN_TRACE_FUNC(PAN_TRACE_GL_RESOURCE);
 
-   afbcp_debug(ctx,
-               "AFBC-P prsrc=%p: Get payload sizes (reads=%u bo_size=%zu, gpu=%s)",
-               prsrc, prsrc->afbcp->nr_consecutive_reads,
-               panfrost_bo_size(prsrc->bo),
-               pan_screen(ctx->base.screen)->afbcp_gpu_payload_sizes ?
-               "true" : "false");
+   mesa_logd("AFBC-P prsrc=%p: Get payload sizes (reads=%u bo_size=%zu, gpu=%s)",
+             prsrc, prsrc->afbcp->nr_consecutive_reads,
+             panfrost_bo_size(prsrc->bo),
+             pan_screen(ctx->base.screen)->afbcp_gpu_payload_sizes ?
+             "true" : "false");
 
    struct panfrost_screen *screen = pan_screen(ctx->base.screen);
    struct panfrost_device *dev = pan_device(ctx->base.screen);
@@ -2062,10 +2061,9 @@ pan_resource_afbcp_get_payload_offsets(struct panfrost_context *ctx,
 {
    PAN_TRACE_FUNC(PAN_TRACE_GL_RESOURCE);
 
-   afbcp_debug(ctx,
-               "AFBC-P prsrc=%p: Get payload offsets (reads=%u bo_size=%zu)",
-               prsrc, prsrc->afbcp->nr_consecutive_reads,
-               panfrost_bo_size(prsrc->bo));
+   mesa_logd("AFBC-P prsrc=%p: Get payload offsets (reads=%u bo_size=%zu)",
+             prsrc, prsrc->afbcp->nr_consecutive_reads,
+             panfrost_bo_size(prsrc->bo));
 
    struct panfrost_device *dev = pan_device(ctx->base.screen);
    uint64_t modifier = prsrc->modifier;
@@ -2119,9 +2117,9 @@ pan_resource_afbcp_pack(struct panfrost_context *ctx,
 {
    PAN_TRACE_FUNC(PAN_TRACE_GL_RESOURCE);
 
-   afbcp_debug(ctx, "AFBC-P prsrc=%p: Pack (reads=%u bo_size=%zu ratio=%.2f)",
-               prsrc, prsrc->afbcp->nr_consecutive_reads,
-               panfrost_bo_size(prsrc->bo), prsrc->afbcp->ratio);
+   mesa_logd("AFBC-P prsrc=%p: Pack (reads=%u bo_size=%zu ratio=%.2f)", prsrc,
+             prsrc->afbcp->nr_consecutive_reads, panfrost_bo_size(prsrc->bo),
+             prsrc->afbcp->ratio);
 
    struct panfrost_screen *screen = pan_screen(ctx->base.screen);
    struct panfrost_device *dev = pan_device(ctx->base.screen);
@@ -2166,10 +2164,9 @@ pan_resource_afbcp_commit(struct panfrost_context *ctx,
 {
    PAN_TRACE_FUNC(PAN_TRACE_GL_RESOURCE);
 
-   afbcp_debug(ctx,
-               "AFBC-P prsrc=%p: Commit (reads=%u bo_size=%zu ratio=%.2f)",
-               prsrc, prsrc->afbcp->nr_consecutive_reads,
-               panfrost_bo_size(prsrc->bo), prsrc->afbcp->ratio);
+   mesa_logd("AFBC-P prsrc=%p: Commit (reads=%u bo_size=%zu ratio=%.2f)",
+             prsrc, prsrc->afbcp->nr_consecutive_reads,
+             panfrost_bo_size(prsrc->bo), prsrc->afbcp->ratio);
 
    uint64_t modifier = prsrc->modifier & ~AFBC_FORMAT_MOD_SPARSE;
    assert(!panfrost_is_emulated_mod(modifier));
diff --git a/src/gallium/drivers/panfrost/pan_screen.h b/src/gallium/drivers/panfrost/pan_screen.h
index b04c607d99f1..76902c84dbb5 100644
--- a/src/gallium/drivers/panfrost/pan_screen.h
+++ b/src/gallium/drivers/panfrost/pan_screen.h
@@ -202,11 +202,4 @@ void panfrost_cmdstream_screen_init_v14(struct panfrost_screen *screen);
       util_debug_message(&ctx->base.debug, PERF_INFO, __VA_ARGS__);            \
    } while (0)
 
-#define afbcp_debug(ctx, ...)                                                  \
-   do {                                                                        \
-      if (unlikely(pan_device((ctx)->base.screen)->debug & PAN_DBG_FORCE_PACK)) \
-         mesa_logw(__VA_ARGS__);                                               \
-      util_debug_message(&ctx->base.debug, INFO, __VA_ARGS__);                 \
-   } while (0)
-
 #endif /* PAN_SCREEN_H */
-- 
GitLab


From 40be52c7d19e448f671e6526d4ee35d9c936327c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lo=C3=AFc=20Molinari?= <loic.molinari@collabora.com>
Date: Thu, 10 Sep 2026 14:15:55 +0200
Subject: [PATCH 2/3] panfrost: Defer resource updates at the end of draw calls
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Resource update is used to let resources know that they're being used
by a draw call. It's currently only used for AFBC packing. The issue
is that it's currently called while a draw call is being emitted and
it prevents the update handler from flushing the current batch. Doing
so results in a crash because the batch is submitted and when the draw
call continues it expects the batch to be in the state it was before
the update.

This commit propose to queue the resources at draw call dependency
handling and update them at the end of the draw call when all the
batches can be safely submitted.

Signed-off-by: Loïc Molinari <loic.molinari@collabora.com>
---
 src/gallium/drivers/panfrost/pan_cmdstream.c |  8 ++++++
 src/gallium/drivers/panfrost/pan_context.h   |  7 +++++
 src/gallium/drivers/panfrost/pan_job.c       | 28 +++++++++++++++++---
 src/gallium/drivers/panfrost/pan_job.h       |  5 ++++
 4 files changed, 44 insertions(+), 4 deletions(-)

diff --git a/src/gallium/drivers/panfrost/pan_cmdstream.c b/src/gallium/drivers/panfrost/pan_cmdstream.c
index 6629934fecdf..7efd95891c39 100644
--- a/src/gallium/drivers/panfrost/pan_cmdstream.c
+++ b/src/gallium/drivers/panfrost/pan_cmdstream.c
@@ -3675,6 +3675,10 @@ panfrost_draw_vbo(struct pipe_context *pipe, const struct pipe_draw_info *info,
    } else {
       panfrost_multi_draw_direct(pipe, info, drawid_offset, draws, num_draws);
    }
+
+   /* Update once the draw call is pushed so that resources can safely push
+    * more calls. */
+   panfrost_update_rsrcs(ctx);
 }
 
 /* Launch grid is the compute equivalent of draw_vbo, so in this routine, we
diff --git a/src/gallium/drivers/panfrost/pan_context.h b/src/gallium/drivers/panfrost/pan_context.h
index 75e72ab6c777..9c5c8da7abc6 100644
--- a/src/gallium/drivers/panfrost/pan_context.h
+++ b/src/gallium/drivers/panfrost/pan_context.h
@@ -36,6 +36,9 @@
 
 #include "pan_csf.h"
 
+/* XXX Figure out correct number */
+#define PAN_MAX_RSRCS_PER_DRAW 1024
+
 #define SET_BIT(lval, bit, cond)                                               \
    if (cond)                                                                   \
       lval |= (bit);                                                           \
@@ -143,6 +146,10 @@ struct panfrost_context {
       BITSET_DECLARE(active, PAN_MAX_BATCHES);
    } batches;
 
+   /* Resources accessed per draw call (read/write). */
+   unsigned nr_rsrcs[2];
+   struct panfrost_resource *rsrcs[2][PAN_MAX_RSRCS_PER_DRAW];
+
    /* Map from resources to panfrost_batches */
    struct hash_table *writers;
 
diff --git a/src/gallium/drivers/panfrost/pan_job.c b/src/gallium/drivers/panfrost/pan_job.c
index 6dfde4e2933d..119abb1bae97 100644
--- a/src/gallium/drivers/panfrost/pan_job.c
+++ b/src/gallium/drivers/panfrost/pan_job.c
@@ -238,6 +238,28 @@ panfrost_get_fresh_batch_for_fbo(struct panfrost_context *ctx,
    return batch;
 }
 
+void
+panfrost_add_rsrc(struct panfrost_context *ctx,
+                  struct panfrost_resource *rsrc, bool writes)
+{
+   assert(ctx->nr_rsrcs[writes] < PAN_MAX_RSRCS_PER_DRAW);
+
+   ctx->rsrcs[writes][ctx->nr_rsrcs[writes]++] = rsrc;
+}
+
+void
+panfrost_update_rsrcs(struct panfrost_context *ctx)
+{
+   for (int i = 0; i < ctx->nr_rsrcs[0]; i++)
+      pan_resource_update_access(ctx, ctx->rsrcs[0][i], false);
+
+   for (int i = 0; i < ctx->nr_rsrcs[1]; i++)
+      pan_resource_update_access(ctx, ctx->rsrcs[1][i], true);
+
+   ctx->nr_rsrcs[0] = 0;
+   ctx->nr_rsrcs[1] = 0;
+}
+
 static bool panfrost_batch_uses_resource(struct panfrost_batch *batch,
                                          struct panfrost_resource *rsrc);
 
@@ -280,6 +302,8 @@ panfrost_batch_update_access(struct panfrost_batch *batch,
             panfrost_batch_submit(ctx, other_batch);
       }
    }
+
+   panfrost_add_rsrc(ctx, rsrc, writes);
 }
 
 static pan_bo_access *
@@ -366,8 +390,6 @@ panfrost_batch_read_rsrc(struct panfrost_batch *batch,
 {
    uint32_t access = PAN_BO_ACCESS_READ | panfrost_access_for_stage(stage);
 
-   pan_resource_update_access(batch->ctx, rsrc, false);
-
    panfrost_batch_add_bo_old(batch, rsrc->bo, access);
 
    if (rsrc->separate_stencil)
@@ -385,8 +407,6 @@ panfrost_batch_write_rsrc(struct panfrost_batch *batch,
 {
    uint32_t access = PAN_BO_ACCESS_WRITE | panfrost_access_for_stage(stage);
 
-   pan_resource_update_access(batch->ctx, rsrc, true);
-
    panfrost_batch_add_bo_old(batch, rsrc->bo, access);
 
    if (rsrc->separate_stencil)
diff --git a/src/gallium/drivers/panfrost/pan_job.h b/src/gallium/drivers/panfrost/pan_job.h
index 4e47406ac119..ba67cb97fe2e 100644
--- a/src/gallium/drivers/panfrost/pan_job.h
+++ b/src/gallium/drivers/panfrost/pan_job.h
@@ -254,4 +254,9 @@ panfrost_has_fragment_job(struct panfrost_batch *batch)
    return batch->draw_count > 0 || batch->clear;
 }
 
+void panfrost_add_rsrc(struct panfrost_context *ctx,
+                       struct panfrost_resource *rsrc, bool writes);
+
+void panfrost_update_rsrcs(struct panfrost_context *ctx);
+
 #endif
-- 
GitLab


From 039911e2e4a13a3191c8f8fe92aa9f78770f62af Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lo=C3=AFc=20Molinari?= <loic.molinari@collabora.com>
Date: Thu, 10 Sep 2026 14:40:36 +0200
Subject: [PATCH 3/3] panfrost: Flush AFBC-P compute jobs
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

We can safely flush the AFBC-P compute jobs now that update is called
outside of a draw call.

Signed-off-by: Loïc Molinari <loic.molinari@collabora.com>
---
 src/gallium/drivers/panfrost/pan_resource.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/gallium/drivers/panfrost/pan_resource.c b/src/gallium/drivers/panfrost/pan_resource.c
index dab9f3681f62..461a05886f89 100644
--- a/src/gallium/drivers/panfrost/pan_resource.c
+++ b/src/gallium/drivers/panfrost/pan_resource.c
@@ -2004,6 +2004,8 @@ pan_resource_afbcp_get_payload_sizes(struct panfrost_context *ctx,
       screen->vtbl.afbc_size(batch, prsrc, prsrc->afbcp->layout_bo,
                              prsrc->afbcp->layout_offsets[level], level);
 
+   panfrost_flush_batches_accessing_rsrc(ctx, prsrc,
+                                         "AFBC-P payload sizes compute job");
    prsrc->afbcp->skip_access_updates = false;
 
    return true;
@@ -2152,6 +2154,8 @@ pan_resource_afbcp_pack(struct panfrost_context *ctx,
                              prsrc->afbcp->layout_bo,
                              prsrc->afbcp->layout_offsets[level], level);
 
+   panfrost_flush_batches_accessing_rsrc(ctx, prsrc,
+                                         "AFBC-P packing compute job");
    prsrc->afbcp->skip_access_updates = false;
 
    return true;
-- 
GitLab

